### PR TITLE
Gan validation

### PR DESCRIPTION
**Pull request by Nash2325138**
## Why do we need this PR?
* Add FID score to evaluate GAN training
image

## How is it implemented?
* Add metric class `FIDScoreOnline` and `FIDScoreOffline` utilizing submodule `fid_score`
  * The output numbers will have slightly difference (<1%) derived from missing precision of saved images (0\~1 float -> 0\~255 integer)
  * `FIDScoreOnline` is recommended to use

## Other Changes
* Add `forward_scenario`
* Update GAN training config: turn on spectral norm and change monitored metric to fid score
* Tensorboard: plot distribution of `x` and `G(x)`
* Disable GAN related losses calcualteion on the validation set.

### CODE REVIEW

File: src/utils/util.py [hunk 1]
```diff
@@ -42,3 +47,44 @@ def one_hot_embedding(labels, num_classes):
     """
     y = torch.eye(num_classes)
     return y[labels]
+
+
+class UnNormalize(object):
```
Comment (amjltc295) [inline src/utils/util.py]: Why is it called UnNormailze?

File: src/model/metric.py [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
 import torch
 from abc import abstractmethod
+import tempfile
+
+import numpy as np
+from torchvision import transforms
+
+from utils.util import UnNormalize, lib_path, import_given_path
+
+inception = import_given_path("inception", os.path.join(lib_path, 'pytorch_fid/inception.py'))
```
Comment (amjltc295) [inline src/model/metric.py]: Why do we need to import the lib like this?

File: src/model/metric.py [hunk 3]
```diff
@@ -47,3 +57,93 @@ def update(self, data, output):
...
+
+class FIDScoreOnline(BaseMetric):
+    """
+    Module calculating FID score online (store inception activation in memory)
+    """
+
+    def __init__(self, output_key, target_key, unnorm_mean=(0.5,), unnorm_std=(0.5,), nickname="FID_InceptionV3"):
+        super().__init__(output_key, target_key, nickname)
+        self._unNormalizer = UnNormalize(unnorm_mean, unnorm_mean)
+        block_idx = inception.InceptionV3.BLOCK_INDEX_BY_DIM[2048]
```
Comment (amjltc295) [inline src/model/metric.py]: Is FID bound with InceptionV3?

File: src/utils/util.py (hunk 1 above)
Comment (Nash2325138) [reply src/utils/util.py]: It's the inverse operation of Normailze

File: src/model/metric.py (hunk 2 above)
Comment (Nash2325138) [reply src/model/metric.py]: Because I don't want to change `sys.path` directly

File: src/model/metric.py (hunk 3 above)
Comment (Nash2325138) [reply src/model/metric.py]: No. Are you thinking that the model should be constructed outside `FIDScoreOnline`?

File: src/utils/util.py (hunk 1 above)
Comment (amjltc295) [reply src/utils/util.py]: Unnormalize?

File: src/model/metric.py (hunk 3 above)
Comment (amjltc295) [reply src/model/metric.py]: It should probably be configurable or named as InceptionV3FIDScoreOnline?

File: src/utils/util.py (hunk 1 above)
Comment (Nash2325138) [reply src/utils/util.py]: I'll just call it `InverseNormalize`

File: src/model/metric.py (hunk 3 above)
Comment (Nash2325138) [reply src/model/metric.py]: I've made it an abstract class such that it supports different backbone
